docs/rfcs/20220209_limit_connections_rate.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: Ability to limit connection creation rate per shard
+- Status: in-progress
+- Start Date: 2022-02-09
+- Authors: Vadim Plakhtinskiy
+- Issue: 
+
+# Executive Summary
+
+- Set a limit on the rate with which the shard accepts new connections
+- Set a limit on connection creation rate per IP
+
+
+Similar to KIP-612, adapted to Redpanda's thread per core model
+
+## What is being proposed
+
+New configuration properties (For start we choose global config for connection rate):
+- kafka_max_connection_creation_rate (optional integer, default unset)
+- kafka_max_connection_creation_rate_ovverides (string, default "")
+
+
+When a connection is accepted which exceeds one of the configured
+bounds, it is dropped before reading or writing anything to the socket.
+
+
+## Motivation
+
+* Avoid hitting system resource limits (for example CPU)
+* Avoid geting connections storms
+* Allow broker doing useful work instead of handling high connection creation rate
+
+## How
+
+If both configuration properties are unset, then no action is taken: this
+preserves the existing unlimited behaviour.
+
+We will track conection rate on each shard independently (without cross-core communication).
+
+We will store semaphore to count connection in current second. For ovverides we will store hash table
+from ip -> semaphore.
+
+For new connection we will do:
+1. Update semaphore to limit if need it
```

Comment:
Can you give more detail on the algorithm for updating the semaphore?

If we are only updating the semaphore when handling a new connection, and a connection goes into the wait state, but there are no subsequent connections, who wakes him up?  For handling that case there probably needs to be a timer, who is only activated once some fiber is waiting.

Reply:
I think if we do not block connection updating process is easy. If we see that second changes. we should add max rate limit to semaphore for new second

But with blocking connection it is a little bit harder